doc: boards: extensions: introduce zephyr:board role and directive

kartben opened the pull request:
https://builds.zephyrproject.io/zephyr/pr/79994/docs/boards/seeed/wio_terminal/doc/index.html
https://builds.zephyrproject.io/zephyr/pr/79994/docs/contribute/documentation/guidelines.html#boards

A new `zephyr:board::` Sphinx directive allows to flag a documentation page as being the documentation for a specific board, allowing to auto-populate some of the contents, ex. by adding a board overview *à la Wikipedia*, and later things like supported HW features, a ready-to-copy-paste example of how to build Hello World, etc.

<img width="1331" alt="image" src="https://github.com/user-attachments/assets/204ba606-a0e6-4610-bd83-1f2e0ed916a7">

A corresponding `:zephyr:board:` role allows to link to a board doc page.

kartben committed:
doc: boards: extensions: store board_catalog in Sphinx domain

Load the board catalog only once at build-inited time so that
it's cached and enabled things like boards, vendors, and socs
to eventually become first class Sphinx objects.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: boards: extensions: store "raw" image path

Small change to the way image are stored in the catalog so that we know
where the source image is rather than anticipating on the final path
after build has happened. This allows to potentially reference the
source image from e.g. directives.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: boards: extensions: remove unused import

Make all Zephyr directives inherit from SphinxDirective

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: boards: extensions: add board name to catalog for convenience

This makes dictionary entries from the boards catalog more
self-contained as the name of the board is now included in the entry.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: boards: extensions: introduce zephyr:board role and directive

A new zephyr:board:: Sphinx directive allows  to flag a documentation
page as being the documentation for a specific board, allowing to
auto-populate some of the contents, ex. by adding a board overview a la
Wikipedia, and later things like supported HW features, etc.

A corresponding :zephyr:board: role allows to link to a board doc page.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

fabiobaltieri commented on doc/_extensions/zephyr/domain/__init__.py:
```diff
@@ -213,6 +219,64 @@ def convert_node(self, node):
+
+            if node["image"] is not None:
+                figure = nodes.figure()
+                figure += nodes.image(uri=f"/{node['image']}")
```
cool stuff, should the small image be linked to the full size image? we don't really do high res but looking at the page it's tempting to click (not even sure this is the right place to be honest, I'm talking about the board overview card in the board page https://builds.zephyrproject.io/zephyr/pr/79994/docs/boards/seeed/wio_terminal/doc/index.html)

kartben replied:
Ya it's the right place :) and yes, great suggestion

kartben replied:
> which I think might actually be a bug in RTD theme, btw, so I'll try to submit a fix upstream as well

https://github.com/readthedocs/sphinx_rtd_theme/pull/1622

sylvioalves commented on doc/templates/board.tmpl:
```diff
@@ -1,20 +1,15 @@
 
-[Board Name]
-#############
+.. In order for the board documentation page to show properly in the
```
@kartben overall LGTM, I would just improve this text to:
"To ensure the board documentation page displays correctly, it is highly recommended to include a picture of the board."

gmarull commented:
@kartben unrelated to this PR, but, could we make header less cluttered?

![image](https://github.com/user-attachments/assets/7ee8a02f-1a2b-4e4e-b27b-a0ecf90f541e)

pdgendt commented:
> @kartben unrelated to this PR, but, could we make header less cluttered?
> 
> ![image](https://private-user-images.githubusercontent.com/25011557/377771905-7ee8a02f-1a2b-4e4e-b27b-a0ecf90f541e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyMzgxMzIsIm5iZiI6MTcyOTIzNzgzMiwicGF0aCI6Ii8yNTAxMTU1Ny8zNzc3NzE5MDUtN2VlOGEwMmYtMWEyYi00ZTRlLWIyN2ItYTBlY2Y5MGY1NDFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDA3NTAzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3YTM3ZTc1YmU5M2RkOWMxNDMzYTgyNDk2MWQ1YzJjOTFiN2UxN2FhYTM3NmZmZjU5MGI4NzRkM2IxM2JlOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2vtGlQZoV5A2G966E1SGYSvYUs9RdkhlwurRC8urtsU)

The "Open on github" and "Report an issue with this page" could simply be icons IMO, with the text in a tooltip. Of course accessibility needs to be verified.

kartben commented:
> @kartben unrelated to this PR, but, could we make header less cluttered?
> 
> ![image](https://private-user-images.githubusercontent.com/25011557/377771905-7ee8a02f-1a2b-4e4e-b27b-a0ecf90f541e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyNDA4ODQsIm5iZiI6MTcyOTI0MDU4NCwicGF0aCI6Ii8yNTAxMTU1Ny8zNzc3NzE5MDUtN2VlOGEwMmYtMWEyYi00ZTRlLWIyN2ItYTBlY2Y5MGY1NDFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDA4MzYyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2N2JjODhlZWJlZjljNzgxOTUxNjkwM2FkZjQ1ODMxOTU1YTAxOTc1NmZhYmM2ZjY2ZTExYTE4ZDc4ZWJjZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.d05B6C0CVDu6ZWWQhOcmG8_kgNabpq3tuaQKcTurqFQ)

Agreed -- I too found it particularly striking when making the screen capture :) Will go for text-less version of the GH buttons and will try to find a way to have ellipses in the breadcrumb

carlescufi reviewed (dismissed):
Really nice stuff!

kartben committed:
doc: scripts: boards: don't init pykwalify logging

This might be a leftover from a previous version but this is basically
silencing all Sphinx logs when generating the catalog from Sphinx -- not
good!

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
Just pushed a fix for a nasty bug that was causing Sphinx logs to be silenced - whoopsie.

fabiobaltieri reviewed (dismissed):
Great stuff can't wait to see how this renders in the pdf.

kartben committed:
doc: release: 3.7: sphinxlint fixes

Fixes typos in Sphinx roles

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
> Great stuff can't wait to see how this renders in the pdf.

We don't include boards in the PDF 😝

kartben commented:
Had to push one more commit to clean up 3.7 release notes re: Sphinx lint.
Just like code samples, boards are now referenced using a dedicated role and not the generic `:ref:` anymore so this requires updates tree-wide, and I deliberately hadn't made a pass on past release notes regarding Sphinx linting as I didn't want to "touch" them unnecessarily (but since the lint errors were actual bugs causing bad rendering, I probably should have...)

kartben committed:
boards: seeed: adopt new zephyr:board directive and role

This updates the documentation of all the Seeed Studio boards to use
the new `zephyr:board::` directive.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

gmarull reviewed (approved):
LGTM. It would be nice to start a mass cleanup of board pages. They contain tons of redundant information that just increases doc build time.

kartben commented:
> LGTM. It would be nice to start a mass cleanup of board pages. They contain tons of redundant information that just increases doc build time.

@gmarull this will be incremental and take a lot of time but... yes, definitely -- e.g. https://github.com/zephyrproject-rtos/zephyr/pull/80048

MaureenHelm commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -834,15 +834,15 @@ Drivers and Sensors
   * Added support for Ambiq Apollo3 series.
   * Added support for multiple instances of the SPI NOR driver (spi_nor.c).
   * Added preliminary support for non-erase devices with introduction of
-    device capabilities to c:struct:`flash_parameters` and the utility function
```
Kind of weird to be modifying release notes of an already finalized release.

kartben replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/79994#issuecomment-2423233033 for some context
